### PR TITLE
Importing using rocks

### PR DESCRIPTION
**Pull request by itaiad200**

### CODE REVIEW

File: onboard/rocks_catalog_actions.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	commitID, err := c.entryCatalog.CommitExistingMetaRange(ctx, c.repoID, catalog.DefaultImportBranchName, *c.createdMetaRangeID, c.committer, commitMsg, graveler.Metadata(metadata))
+	if err != nil {
+		return nil, fmt.Errorf("creating commit from existing metarange %s: %w", *c.createdMetaRangeID, err)
+	}
+
+	commit, err := c.entryCatalog.GetCommit(ctx, c.repoID, commitID)
+	if err != nil {
+		return nil, fmt.Errorf("getting current commit %s: %w", commitID, err)
+	}
+	commitLog := &catalog.CommitLog{
```
Comment (johnnyaug) [inline onboard/rocks_catalog_actions.go:79]: In the end, this log is just used to get the commit ref.
I would change the RepoActions interface to return a string in the commit method, and just return the commit ref here and in the other implementation.
Will save this translation here.

**Review by johnnyaug — CHANGES_REQUESTED** [review]

File: graveler/committed/manager.go [hunk 2]
```diff
@@ -26,6 +27,24 @@ func (c *committedManager) List(ctx context.Context, ns graveler.StorageNamespac
 	panic("implement me")
 }
 
+func (c *committedManager) WriteMetaRange(ctx context.Context, ns graveler.StorageNamespace, it graveler.ValueIterator) (*graveler.MetaRangeID, error) {
+	writer := c.metaRangeManager.NewWriter(ns)
+	for it.Next() {
+		if err := it.Err(); err != nil {
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply graveler/committed/manager.go]: Done.

File: onboard/import.go [hunk 3]
```diff
@@ -96,10 +96,10 @@ func (s *Importer) diffIterator(ctx context.Context, commit catalog.CommitLog) (
 	return NewDiffIterator(previousObjs, currentObjs), nil
 }
 
-func (s *Importer) Import(ctx context.Context, dryRun bool) (*Stats, error) {
+func (s *Importer) Import(ctx context.Context, dryRun, rocks bool) (*Stats, error) {
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply onboard/import.go]: Done.

File: onboard/inventory_entry_iterator.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
+package onboard
+
+import (
+	"errors"
+
+	"google.golang.org/protobuf/types/known/timestamppb"
+
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply onboard/inventory_entry_iterator.go]: Done.

File: onboard/inventory_entry_iterator.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+		it: it,
+	}
+}
+
+func (e *inventoryEntryIterator) Next() bool {
+	if e.err != nil {
+		return false
+	}
+	hasNext := e.it.Next()
+	if !hasNext {
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply onboard/inventory_entry_iterator.go]: Done.

File: onboard/inventory_entry_iterator.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+		},
+	}
+
+	if v.Obj.LastModified != nil {
+		e.value.LastModified = timestamppb.New(*v.Obj.LastModified)
+	}
+	return true
+}
+
+var ErrIteratorNotSeakable = errors.New("can't seek on inventory iterator")
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply onboard/inventory_entry_iterator.go]: Done.

File: onboard/rocks_catalog_actions.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
+package onboard
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/treeverse/lakefs/graveler"
+
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply onboard/rocks_catalog_actions.go]: Done.

File: catalog/rocks/cataloger.go [hunk 8]
```diff
@@ -40,6 +40,10 @@ func NewCataloger(db db.Database, cfg *config.Config) (catalog.Cataloger, error)
 	}, nil
 }
 
+func (c *cataloger) GetEntryCatalog() *EntryCatalog {
```
Comment (nopcoder) [inline catalog/rocks/cataloger.go:43]: Temporary or I would try to avoid doing this one

File: catalog/rocks/entry_value_iterator_test.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+	it := NewEntryToValueIterator(entryIterator)
+	defer it.Close()
+
+	var values []*graveler.ValueRecord
+	for it.Next() {
+		values = append(values, it.Value())
+	}
+	// verify that what we produced is what we got from the iterator
+	if diff := deep.Equal(values, expectedRecords); diff != nil {
+		t.Fatal("ValueToEntry iterator found diff:", diff)
```
Comment (nopcoder) [inline catalog/rocks/entry_value_iterator_test.go:39]:
```suggestion
		t.Fatal("EntryToValue iterator found diff:", diff)
```

File: graveler/committed/manager.go [hunk 10]
```diff
@@ -26,6 +27,24 @@ func (c *committedManager) List(ctx context.Context, ns graveler.StorageNamespac
...
+			return nil, fmt.Errorf("getting value from iterator: %w", err)
+		}
+		if err := writer.WriteRecord(*it.Value()); err != nil {
+			return nil, fmt.Errorf("writing record: %w", err)
```
Comment (nopcoder) [inline graveler/committed/manager.go:34]: optional: add %s with the key

File: graveler/committed/manager.go [hunk 11]
```diff
@@ -26,6 +27,24 @@ func (c *committedManager) List(ctx context.Context, ns graveler.StorageNamespac
 	panic("implement me")
 }
 
+func (c *committedManager) WriteMetaRange(ctx context.Context, ns graveler.StorageNamespace, it graveler.ValueIterator) (*graveler.MetaRangeID, error) {
+	writer := c.metaRangeManager.NewWriter(ns)
```
Comment (nopcoder) [inline graveler/committed/manager.go:31]: defer Abort in case we return with an error

File: onboard/rocks_catalog_actions.go (hunk 1 above)
Comment (itaiad200) [reply onboard/rocks_catalog_actions.go]: Done.

File: onboard/inventory_entry_iterator.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+		return false
+	}
+	v := e.it.Get()
+
+	e.value = &rocks.EntryRecord{
+		Path: rocks.Path(v.Obj.Key),
+		Entry: &rocks.Entry{
+			Address: v.Obj.PhysicalAddress,
+			Size:    v.Obj.Size,
+			ETag:    []byte(v.Obj.Checksum),
```
Comment (nopcoder) [inline onboard/inventory_entry_iterator.go:41]: Think the original idea was to keep it as bytes - so in case the checksum is in hex we need to encode

File: onboard/inventory_entry_iterator.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+	if v.Obj.LastModified != nil {
+		e.value.LastModified = timestamppb.New(*v.Obj.LastModified)
+	}
+
+	e.progress.Add(1)
+	return true
+}
+
+var ErrIteratorNotSeekable = errors.New("can't seek on inventory iterator")
```
Comment (nopcoder) [inline onboard/inventory_entry_iterator.go:53]: can you move this one up before all 'func's

File: onboard/rocks_catalog_actions.go [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/treeverse/lakefs/catalog"
+	"github.com/treeverse/lakefs/catalog/rocks"
+	"github.com/treeverse/lakefs/cmdutils"
+	"github.com/treeverse/lakefs/graveler"
+	"github.com/treeverse/lakefs/logging"
+)
+
+// RocksCatalogRepoActions is in-charge of importing data to lakeFS with Rocks implementation
+type RocksCatalogRepoActions struct {
```
Comment (nopcoder) [inline onboard/rocks_catalog_actions.go:16]: I assume this one is temporary as the name and the interface were made to abstract catalog functionality.

File: onboard/inventory_entry_iterator.go (hunk 12 above)
Comment (itaiad200) [reply onboard/inventory_entry_iterator.go:48]: Done.

File: onboard/inventory_entry_iterator.go (hunk 13 above)
Comment (itaiad200) [reply onboard/inventory_entry_iterator.go]: Done.

File: onboard/rocks_catalog_actions.go (hunk 14 above)
Comment (itaiad200) [reply onboard/rocks_catalog_actions.go:16]: Yeah but they can't really abstract for the import scenario. We're using the Catalog/cataloger in a whole different way between the two.

File: graveler/committed/manager.go (hunk 10 above)
Comment (nopcoder) [reply graveler/committed/manager.go:36]: good point - I guess we will have to address a wider solution to trace a request which will include this information.

File: go.mod [hunk 15]
```diff
@@ -28,6 +28,7 @@ require (
 	github.com/go-openapi/validate v0.20.0
 	github.com/go-swagger/go-swagger v0.25.0
 	github.com/go-test/deep v1.0.7
+	github.com/gobuffalo/packr/v2 v2.2.0
```
Comment (nopcoder) [inline go.mod:31]: we are using statik

File: onboard/rocks_catalog_actions.go (hunk 14 above)
Comment (nopcoder) [reply onboard/rocks_catalog_actions.go:16]: Guess it will be part of the pulling out the MVCC implementation

File: graveler/committed/manager.go [hunk 16]
```diff
@@ -26,6 +27,26 @@ func (c *committedManager) List(ctx context.Context, ns graveler.StorageNamespac
...
+	defer func() { _ = writer.Abort() }()
```
Comment (nopcoder) [inline graveler/committed/manager.go:32]: just log the failed abort error/warn

File: onboard/inventory_entry_iterator.go (hunk 12 above)
Comment (nopcoder) [reply onboard/inventory_entry_iterator.go:48]: 1. You can remove the assignment of ETag now from line 48
2. Just make sure that import checksum is actually in Hex format and not base64.

File: onboard/inventory_entry_iterator.go [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+		e.value.LastModified = timestamppb.New(*v.Obj.LastModified)
+	}
+
+	e.progress.Add(1)
+	return true
+}
+
+func (e *inventoryEntryIterator) SeekGE(_ rocks.Path) {
+	e.err = ErrIteratorNotSeekable
+}
```
Comment (nopcoder) [inline onboard/inventory_entry_iterator.go:70]: e.value = nil or condition the reutrn of value in Value() with err check

**Review by nopcoder — COMMENTED** [review]
<details>
<summary>Details</summary>



</details>

File: .gitignore [hunk 18]
```diff
@@ -75,5 +75,6 @@ terraform.rc
 /graveler/committed/mock/
 /graveler/sstable/mock/
 /pyramid/mock/*
+/onboard/mock/*
```
Comment (nopcoder) [inline .gitignore:78]: can you remove the '*' from both lines - the directory is optional

File: catalog/rocks/entry_value_iterator_test.go [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+	entryIterator := NewFakeEntryIterator(entryRecords)
+	it := NewEntryToValueIterator(entryIterator)
+	defer it.Close()
+
+	var values []*graveler.ValueRecord
+	for it.Next() {
+		values = append(values, it.Value())
+	}
+	// verify that what we produced is what we got from the iterator
+	if diff := deep.Equal(values, expectedRecords); diff != nil {
```
Comment (nopcoder) [inline catalog/rocks/entry_value_iterator_test.go:38]: check that the iterator ended without an error
```go
if err := it.Err(); err != nil {
    t.Fatal("Iterator ended with error", err)
}
```

File: onboard/rocks_catalog_actions_test.go [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/treeverse/lakefs/block"
+	"github.com/treeverse/lakefs/onboard"
+
+	"github.com/treeverse/lakefs/logging"
+
+	"github.com/golang/mock/gomock"
+	"github.com/treeverse/lakefs/onboard/mock"
```
Comment (nopcoder) [inline onboard/rocks_catalog_actions_test.go:20]: remove whitespace

File: onboard/rocks_catalog_actions_test.go [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+			{
+				Bucket:          "bucket-1",
+				Key:             "key-1",
+				Size:            1024,
+				Checksum:        "checksum-1",
+				PhysicalAddress: "/some/path/to/object",
+			},
+		}}
+	validIt = onboard.NewInventoryIterator(innerIt)
+)
```
Comment (nopcoder) [inline onboard/rocks_catalog_actions_test.go:43]: move into the test function or create a function that returns this value.
don't use global variable with state.

File: onboard/utils_test.go [hunk 22]
```diff
@@ -122,9 +122,9 @@ func (m *mockCatalogActions) GetPreviousCommit(_ context.Context) (commit *catal
 	return &catalog.CommitLog{Metadata: metadata}, nil
 }
 
-func (m *mockCatalogActions) Commit(_ context.Context, _ string, metadata catalog.Metadata) (*catalog.CommitLog, error) {
+func (m *mockCatalogActions) Commit(_ context.Context, _ string, metadata catalog.Metadata) (string, error) {
```
Comment (nopcoder) [inline onboard/utils_test.go:125]: question: why did you change the signature of the return type?

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: onboard/rocks_catalog_actions_test.go [hunk 23]
```diff
@@ -0,0 +1,125 @@
+package onboard_test
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/treeverse/lakefs/catalog"
+
```
Comment (johnnyaug) [inline onboard/rocks_catalog_actions_test.go:9]: remove newlines

**Review by johnnyaug — APPROVED** [review]

File: .gitignore (hunk 18 above)
Comment (itaiad200) [reply .gitignore]: Done.

File: catalog/rocks/entry_value_iterator_test.go (hunk 19 above)
Comment (itaiad200) [reply catalog/rocks/entry_value_iterator_test.go:37]: Done.

File: onboard/rocks_catalog_actions_test.go (hunk 23 above)
Comment (itaiad200) [reply onboard/rocks_catalog_actions_test.go]: Done.

File: onboard/rocks_catalog_actions_test.go (hunk 20 above)
Comment (itaiad200) [reply onboard/rocks_catalog_actions_test.go]: Done.

File: onboard/utils_test.go (hunk 22 above)
Comment (itaiad200) [reply onboard/utils_test.go:125]: It's redundant. Existing and new code only use the ID.

File: onboard/rocks_catalog_actions_test.go (hunk 21 above)
Comment (itaiad200) [reply onboard/rocks_catalog_actions_test.go]: Done.

File: catalog/rocks/entry_value_iterator_test.go [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+		t.Fatal("EntryToValue iterator found diff:", diff)
+	}
+
+	it.Close()
+	if err := it.Err(); err != nil {
```
Comment (nopcoder) [inline catalog/rocks/entry_value_iterator_test.go:42]: Move to line 35.5 - after the Next() loop. At this point, the iterator is closed.

File: onboard/inventory_entry_iterator.go [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+	"github.com/treeverse/lakefs/catalog/rocks"
+)
+
+type inventoryEntryIterator struct {
+	it    *InventoryIterator
+	value *rocks.EntryRecord
+	err   error
+}
+
+func NewValueToEntryIterator(it *InventoryIterator) rocks.EntryIterator {
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply onboard/inventory_entry_iterator.go]: The assertion is needed. It's not that the code can't handle the interface, it's that the passing the diffIt is a huge mistake.

Once MVCC code is gone, the code can be a lot simpler. For instance, `diffIt` can be deleted.

**Review by nopcoder — APPROVED** [review]

File: catalog/rocks/entry_value_iterator_test.go (hunk 24 above)
Comment (itaiad200) [reply catalog/rocks/entry_value_iterator_test.go:42]: Done.
